profiles/common/modules/features/nexteuropa_newsroom/classes/NewsroomHelper.inc

```diff
@@ -0,0 +1,633 @@
+      }
+      else {
+        $topics = array();
+      }
+
+      $topics = array_merge($primary_topics, $topics);
+
+      cache_set($cache_key, $topics, NEXTEUROPA_NEWSROOM_CACHE_TABLE);
+    }
+    return $topics;
+  }
+
+  /**
+   * Active menu for newsroom items.
+   *
+   * Returns wich menu item should be active for a newsroom item
+   * based on the assigned topics.
+   */
+  public static function fetchActiveMenu($topic_id) {
+    $cache_key = 'fetch_active_menu' . $topic_id;
+    $cache = cache_get($cache_key, NEXTEUROPA_NEWSROOM_CACHE_TABLE);
+    if ($cache) {
+      $path = $cache->data;
+    }
+    else {
+      $path = FALSE;
+
+      // Check if they are added to menu.
+      $menu_items = array();
+      // Fetch entities associated to topic using the special field.
+      $topic_entities = NewsroomHelper::getNodesAssociatedToTopics($topic_id);
+
+      if (count($topic_entities) > 0) {
+        foreach ($topic_entities as $entity) {
+          // Could use entity_uri() but we would have to load the full entity...
+          $menu_link_items = NewsroomHelper::getMenuByPath(str_replace('_', '/', $entity['entity_type']) . '/' . $entity['entity_id']);
+
+          if ($menu_link_items !== FALSE) {
+            foreach ($menu_link_items as $menu_item) {
+              $menu_items[] = $menu_item;
+            }
+          }
+        }
+      }
+
+      if (count($menu_items) > 0) {
+        // Favor terms over nodes and lower entity ID in case we have multiple.
+        usort($menu_items, function ($a, $b) {
+          if ($a['router_path'] == 'taxonomy/term/%' && $b['router_path'] == 'node/%') {
+            return -1;
+          }
+          elseif ($a['router_path'] == 'node/%' && $b['router_path'] == 'taxonomy/term/%') {
+            return 1;
+          }
+          else {
+            if ($a['plid'] == $b['plid']) {
+              if ($a['weight'] == $b['weight']) {
+                if ($a['mlid'] == $b['mlid']) {
+                  return 0;
+                }
+                else {
+                  return $a['mlid'] < $b['mlid'] ? -1 : 1;
+                }
+              }
+              else {
+                return $a['weight'] < $b['weight'] ? -1 : 1;
+              }
+            }
+            else {
+              return $a['plid'] < $b['plid'] ? -1 : 1;
+            }
+          }
+        });
+
+        $menu_item_link = array_shift($menu_items);
+        $path = $menu_item_link['link_path'];
+      }
+
+      cache_set($cache_key, $path, NEXTEUROPA_NEWSROOM_CACHE_TABLE);
+    }
+
+    return $path;
+  }
+
+  /**
+   * Try to resolve each taxonomy term with a corresponding ID to map.
+   *
+   * @param string $field_name
+   *   The mapping field.
+   * @param string $field_value
+   *   The mapping field value.
+   *
+   * @return array
+   *   Taxonomy ids.
+   */
+  public static function getTaxonomyTermFromField($field_name, $field_value, $bundle = NULL) {
```

Comment:
so the original message was to try to find another way to having these data, not exactly the existing of this function is the problem (can be), but the logic inside how it is implemented, especially with the caching solution. Normally this data won't change often that's why i called it 'almost static data', so it is possible to store it in a drupal variable once, build an admin interface for changing it and using variable_get when it needs instead of asking always drupal for creating the result.